### PR TITLE
[qk_clip] add post-optimizer qk clip per step

### PR DESCRIPTION
**Pull request by shuhuayu**
## Summary

- add configurable per-head QK clipping for Kimi MLA attention
- collect each layer's maximum attention logits through FlexAttention auxiliary outputs
- pack all layer/head maxima into one `MAX` all-reduce over the loss mesh per optimizer step
- scale the local Q and K projection weights after the optimizer update while leaving V unchanged
- add CPU and distributed tests covering statistic collection, clipping math, DTensor shards, and cross-rank reduction

## Motivation

Kimi training clips QK attention logits after every optimizer step. The clipping factor must use the maximum logit observed for each query head across all local forwards and all data/context-parallel ranks.

This implementation records per-head maxima during training forwards, combines gradient-accumulation and activation-recomputation observations, and performs one packed collective for all layers and heads. Weight scaling is local after that collective.

## VarlenAttention

This PR supports the FlexAttention backend. FlexAttention can return `max_scores` with shape `[B, N, L]`, which provides the exact pre-softmax statistic required by QK clipping.

The current `torch.nn.attention.varlen.AuxRequest` exposes only `lse`; it does not expose the per-query maximum score. LSE cannot be converted back to the exact maximum, and recomputing `Q @ K.T` outside the fused attention kernel would defeat the purpose of the VarlenAttention backend.

Clean VarlenAttention support requires extending the PyTorch/FlashAttention auxiliary-output contract to expose either:

- per-query maxima with shape `[N, T]`, followed by a reduction over packed tokens, or
- a kernel-reduced per-head maximum with shape `[N]`.

Once that statistic is available, the packed all-reduce and post-optimizer weight-scaling path in this PR can be reused unchanged. Until then, Kimi configurations reject non-FlexAttention backends when QK clipping is enabled.

cc: @drisspg @liangel-02 

## Testing

- `pytest tests/unit_tests/test_qk_clip.py` (`3 passed`)
- `pytest tests/unit_tests/test_compile_regional_inductor.py::TestRegionalInductorBackend` (`5 passed`)
- 2-GPU Kimi debug-model training with `EP=2` for 10 steps
- profiler verification: one packed 96-element FP32 QK-clip all-reduce per step for 6 layers x 16 heads; Q/K scaling afterward is local

### CODE REVIEW

**Comment by shuhuayu** [comment]
confirmed from trace that only one rs is used since all per layer per head max logits are stacked into one single tensor. cc: @weifengpy 
image

**Comment by liangel-02** [comment]
@shuhuayu thanks for flagging, i'll look into varlen gaps for this

File: torchtitan/models/kimi_k2_7/qk_clip.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    if isinstance(param, DTensor):
+        local_param = param.to_local()
+        local_rows = global_shape[0]
+        row_offset = 0
+        for mesh_axis, (mesh_axis_size, placement) in enumerate(
+            zip(param.device_mesh.shape, param.placements, strict=True)
+        ):
+            if type(placement) is Replicate:
+                continue
+            if type(placement) is not Shard or placement.dim % param.ndim != 0:
```
Comment (weifengpy) [inline torchtitan/models/kimi_k2_7/qk_clip.py]: nit: add code comment of TODO here? once supporting FSDP2+TP,  we would have _StridedShard(0) on DP mesh. that would be a good time to extend this part

Comment (weifengpy) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: why `% param.ndim` instead of `placement.dim != 0` ?

Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: this is to accommodate the use of -1, -2 on dimension.

File: torchtitan/models/kimi_k2_7/qk_clip.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                    "dimension 0."
+                )
+            local_rows, local_offset = Shard.local_shard_size_and_offset(
+                local_rows,
+                mesh_axis_size,
+                param.device_mesh.get_local_rank(mesh_axis),
+            )
+            row_offset += local_offset
+    else:
+        local_param = param
```
Comment (weifengpy) [inline torchtitan/models/kimi_k2_7/qk_clip.py]: when can we reach this else branch? is this line for oss titan? model.parameters() are Dtensor at this moment

File: torchtitan/models/kimi_k2_7/qk_clip.py (hunk 1 above)
Comment (weifengpy) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: is Shard(-1) or Shard(-2) possible in the first place? but i don't mind if you are doing normalzation

File: torchtitan/models/kimi_k2_7/qk_clip.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    scales_N: torch.Tensor,
+    *,
+    head_extent: int,
+) -> tuple[int, int]:
+    global_shape = param.shape
+    if len(global_shape) != 2 or global_shape[0] != scales_N.numel() * head_extent:
+        raise ValueError("QK clip scales do not match the MLA weight shape.")
+
+    if isinstance(param, DTensor):
+        local_param = param.to_local()
```
Comment (weifengpy) [inline torchtitan/models/kimi_k2_7/qk_clip.py]: for this DTensor branch, gradient clipping and TP compute implicitly shows torchtitan Dtensor dispatch convention
* DTensor dispatch within a mesh; raw collectives only to bridge across meshes DTensor cannot span (PP, EP)

`dist.all_reduce(max_logits_N, op=dist.ReduceOp.MAX, group=reduction_mesh.get_group())` qualified for raw collectives / hand rolled implementation

`_local_head_range` does not seem to qualify

this is more like a torchittan coding style requirements. the logic itself is correct. cc @tianyu-l if you care about DTensor dispatch vs hand rolled implementation like grad clipping across PP and EP mesh. half of muonclip qualifies hand rolled implemntation

Comment (weifengpy) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: a DTensor dispatch approach: https://github.com/pytorch/torchtitan/pull/4252 with that requires a single all-reduce. this is to make my point on `if isinstance(param, DTensor):` branch

File: torchtitan/models/kimi_k2_7/model.py [hunk 4]
```diff
@@ -78,6 +80,7 @@ def update_from_config(
 
     def __init__(self, config: Config):
         super().__init__(config)
+        self.qk_clip_config = config.qk_clip
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k2_7/model.py]: Is this ever set? Can we hardcode it in qk_clip function for now?

File: torchtitan/models/common/attention.py [hunk 5]
```diff
@@ -283,6 +283,12 @@ def __init__(self, config: Config) -> None:
         super().__init__()
         self.kernel_options = config.kernel_options
 
+    def _get_aux_request(self, *, return_lse: bool) -> AuxRequest:
+        return AuxRequest(lse=return_lse)
+
+    def _process_aux(self, aux: Any) -> None:
+        pass
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:278]: non-trivial, add docstring

File: torchtitan/models/common/attention.py
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:1]: could you make sure this doesn't break RL side of FlexAttention
cc @liangel-02 could you help confirm

File: torchtitan/models/kimi_k2_7/qk_clip.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    @dataclass(kw_only=True, slots=True)
+    class Config(FlexAttention.Config):
+        pass
+
+    def __init__(self, config: Config) -> None:
+        super().__init__(config)
+        self.max_attention_logits_N: list[torch.Tensor] = []
+
+    def _get_aux_request(self, *, return_lse: bool) -> AuxRequest:
+        return AuxRequest(lse=return_lse, max_scores=self.training)
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k2_7/qk_clip.py:35]: looks like it should support inference as well
@liangel-02 do we have the plumbing on vllm-side Flex so that RL would just work?

File: torchtitan/models/kimi_k2_7/qk_clip.py (hunk 1 above)
Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: you are right this is kind of defensive since currently kimi and fsdp do not use negative numbers for dim. but the code costs nothing so kept it. Added the TODO for _StridedShard(0), thx for the pointer.

File: torchtitan/models/kimi_k2_7/qk_clip.py (hunk 3 above)
Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: make sense to me, and the dtensor metadata based per head scaling is more extendable for the stridedshard case after we add tp. I refactored it with your approach and credited to you in the pr authorship. thx.

File: torchtitan/models/common/attention.py
Comment (shuhuayu) [reply torchtitan/models/common/attention.py:1]: code path wise RL uses the base `FlexAttention` and should be safe.

File: torchtitan/models/common/attention.py (hunk 5 above)
Comment (shuhuayu) [reply torchtitan/models/common/attention.py:278]: added.

File: torchtitan/models/kimi_k2_7/model.py (hunk 4 above)
Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/model.py]: no, just used the default value. make sense to keep it simple if no `alpha` is set yet. updated.

File: torchtitan/models/kimi_k2_7/qk_clip.py (hunk 2 above)
Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: updated.

**Review by weifengpy — APPROVED** [review]
thanks for addressing all my comments. will let you triage on per-head clipping

File: torchtitan/models/kimi_k2_7/qk_clip.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+def _replicated_scales(scales_N: torch.Tensor, weight: torch.Tensor) -> torch.Tensor:
+    """Represent per-head scales on the same distributed mesh as ``weight``.
+
+    The MAX all-reduce already leaves identical scales on every rank. For a
+    DTensor weight, ``from_local`` records the scales as replicated without
+    communication, allowing DTensor dispatch to align them with sharded heads.
+    Using ``distribute_tensor`` would add an unnecessary broadcast.
+    """
+    if not isinstance(weight, DTensor):
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k2_7/qk_clip.py]: Does it work with spmd_types?

I would prefer not adding DTensor (aka `partial_dtensor`) backend support -- doesn't have to delete DTensor in this PR.

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/kimi_k2_7/qk_clip.py (hunk 7 above)
Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: it works with `spmd_types`, and persistent parameters are still dtensor under fsdp2 with spmd_types.

Comment (tianyu-l) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: oh, that's right, but it should always be DTensor, no exception, so no if-else? Even with PP only, we use fully_shard with dp degree 1

Comment (shuhuayu) [reply torchtitan/models/kimi_k2_7/qk_clip.py]: you are right, this is only used in the cpu unit test. i will remove the dtensor check and update the test.

File: torchtitan/models/common/attention.py
Comment (liangel-02) [reply torchtitan/models/common/attention.py:1]: yes i think this should be fine

File: torchtitan/models/kimi_k2_7/qk_clip.py (hunk 6 above)
Comment (liangel-02) [reply torchtitan/models/kimi_k2_7/qk_clip.py:35]: yes i think so
